Smash string similarity

Pelanglene opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added new string similarity functions smashSimilarity and smashSimilarityUTF8 that provide intelligent word-based string comparison with gap handling and subsequence matching.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

You can read more about SMASH here:
[P4104 Tang.pdf](https://github.com/user-attachments/files/19865336/P4104.Tang.pdf)

Commit:
sync

Commit:
fix

Commit:
add tests

Commit:
add docs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=79461) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=79461) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: alexey-milovidov<br/>:x: Pelanglene<br/><hr/>**Pelanglene** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=79461) it.</sub>

Commit:
fix test

vdimir commented on tests/queries/0_stateless/03370_smash_similarity.sql:
```diff
@@ -0,0 +1,42 @@
+
+-- Comparison with other string distance functions
+SELECT smashSimilarity('hello world', 'hello there') AS smash;
+
```
Consider adding cases that are close to the limit or hit the limit, for example:

```SQL
SELECT smashSimilarity(repeat('a', 500000), repeat('a', 500000));
SELECT smashSimilarity(repeat('a', 500000), repeat('a', 499999) || 'b');

SELECT smashSimilarity(repeat('ab', 1000), repeat('ba', 1000));
```

Pelanglene replied:
Is there an option to continue executing the tests when there is an exception in the test? I'm trying to add tests that hit the limits, I get an exception in first test that hits the limits and the test execution stops there.

vdimir commented on src/Functions/FunctionsStringDistance.cpp:
```diff
@@ -469,6 +471,299 @@ struct ByteJaroWinklerSimilarityImpl
+        }
+        
+        /// Fill in the matrices
+        for (size_t i = 1; i <= haystack_size; ++i)
+        {
+            for (size_t j = 1; j <= needle_size; ++j)
+            {
+                bool match;
+                if constexpr (is_utf8)
+                    match = haystack_codepoints[i-1] == needle_codepoints[j-1];
+                else
+                    match = haystack[i-1] == needle[j-1];
```
Consider adding more detailed comments for the dynamic programming matrices to explain the algorithm's approach

Pelanglene replied:
Check if it is clear now please

vdimir commented on src/Functions/FunctionsStringDistance.cpp:
```diff
@@ -469,6 +471,299 @@ struct ByteJaroWinklerSimilarityImpl
+        if constexpr (utf8_enabled)
+        {
+            size_t code_point_idx = pos / sizeof(UInt32);
+            return code_point_idx < codepoints.size() && codepoints[code_point_idx] == ' ';
+        }
+        else
+        {
+            return str[pos] == ' ';
```
Perhaps it's also worth handling different whitespace chars, such as `\t` `\n` and so on

vdimir reviewed (commented):
Thanks for the contribution!
I've taken a brief look, but didn't read the algorithm itself carefully yet. I will try to review it more thoroughly later. Please see my initial comments for now.

Commit:
better string size limit check

Pelanglene commented on src/Functions/FunctionsStringDistance.cpp:
```diff
@@ -469,6 +471,299 @@ struct ByteJaroWinklerSimilarityImpl
+{
+    using ResultType = Float64;
+    static constexpr size_t max_string_size = 1000000; // 1MB safety limit
+    static constexpr double big_value = 1e10; // Large but not infinity to avoid overflow
```
It was an int64_t before. Now when its calculated in double I can use `std::numeric_limits<double>::infinity()`, thanks!

Commit:
affine gap docs&tests

Commit:
fix docs

Pelanglene commented on src/Functions/FunctionsStringDistance.cpp:
```diff
@@ -469,6 +471,299 @@ struct ByteJaroWinklerSimilarityImpl
+        const char* __restrict haystack, size_t haystack_size,
+        const char* __restrict needle, size_t needle_size)
+    {
+        if (haystack_size > max_string_size || needle_size > max_string_size)
```
Yes, it's a good point, I added this check. 
I also added time limit check in algorithm. I'm not sure it is a good solution, but I have no idea how to do it differently, because it is difficult to predict whether the algorithm will be executed in a reasonable time

Commit:
detailed comments

Commit:
better isSpace

Commit:
fix

Pelanglene commented on src/Functions/FunctionsStringDistance.cpp:
```diff
@@ -469,6 +471,299 @@ struct ByteJaroWinklerSimilarityImpl
+            if (long_str.size() < short_str.size())
+                std::swap(long_str, short_str);
+            
+            return processStrings(long_str, short_str, haystack_codepoints);
```
Looks like you are right, fixed

Commit:
fix

alexey-milovidov committed:
Delete aspell-dict.txt (removed in master)

`aspell-dict.txt` was deleted in master along with the `aspell` and
`codespell` style checks. Removing the file here so the PR can merge
cleanly without a modify/delete conflict.
